jax/numpy/lax_numpy.py

```diff
@@ -1570,6 +1570,94 @@ def _check_no_padding(axis_padding, mode):
     raise ValueError(msg.format(mode))
 
 
+def _pad_constant(array, pad_width, constant_values):
+  nd = ndim(array)
+  constant_values = broadcast_to(asarray(constant_values), (nd, 2))
+  constant_values = lax.convert_element_type(constant_values, array.dtype)
+  for i in range(nd):
+    widths = [(0, 0, 0)] * nd
+    widths[i] = (pad_width[i, 0], 0, 0)
+    array = lax.pad(array, constant_values[i, 0], widths)
+    widths[i] = (0, pad_width[i, 1], 0)
+    array = lax.pad(array, constant_values[i, 1], widths)
+  return array
+
+
+def _pad_wrap(array, pad_width):
+  for i in range(ndim(array)):
+    if array.shape[i] == 0:
+      _check_no_padding(pad_width[i], "wrap")
+      continue
+    size = array.shape[i]
+    repeats, (left_remainder, right_remainder) = _divmod(pad_width[i], size)
+    total_repeats = repeats.sum() + 1
+    parts = []
+    if left_remainder:
+      parts += [lax.slice_in_dim(array, size - left_remainder, size, axis=i)]
+    parts += total_repeats * [array]
+    if right_remainder:
+      parts += [lax.slice_in_dim(array, 0, right_remainder, axis=i)]
+    array = lax.concatenate(parts, dimension=i)
+  return array
+
+
+def _pad_symmetric_or_reflect(array, pad_width, mode):
+  assert mode in ("symmetric", "reflect")
+
+  for i in range(ndim(array)):
+    if array.shape[i] == 0:
+      _check_no_padding(pad_width[i], mode)
+      continue
+
+    n = array.shape[i]
+    rarray = lax.rev(array, dimensions=(i,))
+    offset = 1 if (mode == "reflect" and n > 1) else 0
+
+    def build_padding(padding, forward):
+      xs = []
+      delta = n - offset
+      while padding > delta:
+        padding -= delta
+        p = array if forward else rarray
+        xs.append(lax.slice_in_dim(p, offset, n, axis=i))
+        forward = not forward
+      if padding > 0:
+        x = lax.slice_in_dim(array if forward else rarray, offset,
+                             padding + offset, axis=i)
+        xs.append(x)
+      return xs
+
+    parts = reversed(build_padding(pad_width[i, 0], forward=True))
+    parts = [lax.rev(x, dimensions=(i,)) for x in parts]
+    parts += [array]
+    parts += build_padding(pad_width[i, 1], forward=False)
+    array = lax.concatenate(parts, dimension=i)
+  return array
+
+
+def _pad_edge(array, pad_width):
+  nd = ndim(array)
+  for i in range(nd):
+    # TODO: what is this business about?  It we have a zero length axis, then our array
+    # contains no values at all, so this padding operation cannot work---here and for reflect,
+    # symmetric, wrap.
```

Comment:
is this a TODO note or an explanation? :)

For what it's worth, I think you did figure this out! But it probably makes sense to leave this comment on either the first place where this check appears or all of them, not just here.

Reply:
I don't want to check this in ... I placed it here for discussion ... it seems to me that this check is not necessary or even correct.  Am I missing something?

Reply:
If the array shape 0 along an axes, then most of these padding method are not well defined. E.g., consider an 1d array with shape `(0,)`. What does it mean to do "edge" padding with a width of `1` on each side? In principle the result should have shape `(2,)`, but there are no valid edge values to pad with in the first place.

Reply:
Right ... if *any axis* has dim 0, then basically all modes except "constant" are out the window.  Even if the axes where padding is requested (e.g. `i` where `pad_width[i]` is not `(0,0)`) are disjoint from the "zero length" axes, this pad operation is still not viable, because there are no entries in an array with zero volume.

So I guess what I am saying is that this "check" repeated across these pad variants is actually incorrectly formulated.  Do you agree?

Reply:
Suppose the input has shape `(0, 10)` and you're asked to pad with `mode='edge'` by `pad_width=((0, 0), (1, 1))`. This has a valid output, another empty array of shape `(0, 12)`.

Reply:
Ah, indeed.  OK then, let me remove my TODO then.  Is best practice to push a new commit or amend and force push?

Reply:
Either way is fine. Usually I recommend pushing an extra commit since it does a better job of preserving history in pull requests.

Reply:
Great.  Pushed.  Thanks for reviewing!